docs/LANGUAGE.md

```diff
@@ -0,0 +1,767 @@
+
+```opalog
+s :- x > y, y = 41, x = 42
+```
+
+The query result is the same:
+
+```
+s
+# true
+```
+
+Opalog supports [References](#references) to nested documents. For example:
+
+```opalog
+sites :- [{"name": "prod"}, {"name": "smoke1"}, {"name": "dev"}]
+r :- sites[].name = "prod"
+```
+
+The rule "r" above asserts that there exists one document within the "sites"
+document which has the name "prod".
+
+The result:
+
+```
+r
+# true
+```
+
+We can generalize the example above with a rule that defines a set document instead of a boolean document:
+
+```opalog
+sites :- [{"name": "prod"}, {"name": "smoke1"}, {"name": "dev"}]
+q[] = name :- sites[].name = name
+```
+
+When we query for "q" we obtain a set of names:
+
+```
+q
+# ["prod", "smoke1", "dev"]
+```
+
+We can re-write the rule "r" from above to make use of "q". We will call the new rule "p":
+
+```opalog
+p :- "prod" = q[]
+```
+
+The result will be the same:
+
+```
+p
+# true
+```
+
+Rules which have arguments can be queried with input values:
+
+```
+q[] = "smoke2"
+# false
+
+q[] = "dev"
+# true
+```
+
+If you made it this far, congratulations. This section introduced the main
+aspsects of Opalog. The rest of this document provides more detail on various
+aspects of Opalog. Opalog's syntax is defined at the end of this document in
+the [Opalog Grammar](#grammar) section.
+
+## <a name="scalar-values"></a> Scalar Values
+
+Scalar values are the simplest type of term in Opalog. Scalar values can be
+strings, numbers, booleans, or null.
+
+Rules can be defined solely in terms of scalar values. This is useful for
+defining constants that are referenced in multiple places. For example:
+
+```opalog
+greeting   :- "Hello"
+max_height :- 42
+pi         :- 3.14159
+allowed    :- true
+sentinel   :- null
+```
+
+These rules can be queried like any other:
+
+```
+greeting
+# "foo"
+
+max_height
+# 42
+
+pi
+# 3.14159
+
+allowed
+# true
+
+sentinel
+# null
+```
+
+## <a name="composite-values"></a> Composite Values
+
+Composite values define collections. In simple cases, composite values
+can be treated as constants like [Scalar Values](#scalar-values):
+
+```opalog
+cube :- {"width": 3, "height": 4, "depth": 5}
+```
+
+The result:
+
+```
+cube.width
+# 3
+```
+
+Composite values can also be defined in terms of [Variables](#variables) or
+[References](#references). For example:
+
+```opalog
+p[] = x :-
+  foo = 42,
+  bar = false,
+  baz = null,
+  x = {"foo": foo, "bar": [bar, baz]}
+```
+
+The result:
+
+```
+p
+# [{"foo": 42, "bar": [false, null]}]
+```
+
+By defining composite values in terms of variables and references, rules can
+define abstractions over raw data and other rules.
+
+## <a name="variables"></a> Variables
+
+Variables are another kind of term in Opalog. They can appear in both the
+head and body of rules.
+
+Variables appearing in the head of a rule can be thought of as input and
+output of the rule. For example:
```

Comment:
Done. The extra prose is good because it is explicit and then the example can help the reader build up intuition.